src/OrchardCore.Modules/OrchardCore.Media/Shortcodes/AssetUrlShortcodeProvider.cs

```diff
@@ -0,0 +1,116 @@
+            _httpContextAccessor = httpContextAccessor;
+            _options = options.Value;
+        }
+
+        public ValueTask<string> EvaluateAsync(string identifier, Arguments arguments, string content, Context context)
+        {
+            if (!String.Equals(identifier, Identifier, StringComparison.OrdinalIgnoreCase))
+            {
+                return Null;
+            }
+
+            // Handle self closing shortcodes.
+            if (String.IsNullOrEmpty(content))
+            {
+                content = arguments.NamedOrDefault("src");
+                if (String.IsNullOrEmpty(content))
+                {
+                    return AssetUrlShortcode;
+                }
+            }
+
+            if (!content.StartsWith("//", StringComparison.Ordinal) && !content.StartsWith("http", StringComparison.OrdinalIgnoreCase))
+            {
+                // Serve static files from virtual path.
+                if (content.StartsWith("~/", StringComparison.Ordinal))
+                {
+                    content = _httpContextAccessor.HttpContext.Request.PathBase.Add(content[1..]).Value;
+                    if (!String.IsNullOrEmpty(_options.CdnBaseUrl))
+                    {
+                        content = _options.CdnBaseUrl + content;
+                    }
+                }
+                else
+                {
+                    content = _mediaFileStore.MapPathToPublicUrl(content);
+                }
+            }
+
+            if (arguments.Any())
+            {
+                var queryStringParams = new Dictionary<string, string>();
+
+                var width = arguments.Named("width");
+                var height = arguments.Named("height");
+                var mode = arguments.Named("mode");
+                var quality = arguments.Named("quality");
+                var format = arguments.Named("format");
+
+                if (width != null)
+                {
+                    queryStringParams.Add("width", width);
+                }
+
+                if (height != null)
+                {
+                    queryStringParams.Add("height", height);
+                }
+
+                if (mode != null)
+                {
+                    queryStringParams.Add("rmode", mode);
+                }
+
+                if (quality != null)
+                {
+                    queryStringParams.Add("quality", quality);
+                }
+
+                if (format != null)
+                {
+                    queryStringParams.Add("format", format);
+                }
+
+                content = QueryHelpers.AddQueryString(content, queryStringParams);
+            }
+
+            // To sanitize the content, which may be later wrapped in a tag by the writer, we wrap it in an img tag, sanitize it, and remove the img tag.
+            content = "<img src=\"" + content + "\">";
+            content = _htmlSanitizerService.Sanitize(content);
```

Comment:
Uri.EscapeDataString

Reply:
Did not work (loses relativity through the escaping). Have tested post sanitize instead (works fine), so it's safe